rqt_reconfigure: add refresh button

cottsay opened the pull request:
Includes fix from https://github.com/ros-visualization/rqt_common_plugins/pull/297 with additional work to make the _refresh_nodes method work more cleanly.

Thanks,

--scott

ablasdel commented:
@cottsay this all looks good to me. 

I plan to merge this before the next release tomorrow or the next day.

dirk-thomas commented on rqt_reconfigure/src/rqt_reconfigure/treenode_qstditem.py:
```diff
@@ -103,6 +103,11 @@ def __init__(self, *args):
         except IndexError:  # tuple index out of range etc.
                 rospy.logerr('TreenodeQstdItem IndexError')
 
+    def __del__(self):
```
Since Python won't call this in a deterministic manner this should be done in an explicitly called method. How does that relate to the method `clear_dynreconf_client`?

ablasdel commented:
I had planned to merge this until dirk made his comment. @cottsay can we fix this issue so we can merge this?

cottsay commented:
Okay, I updated it to remove the destructor...although I'd like to hear more about why @dirk-thomas doesn't think they should be used here.

dirk-thomas commented:
I don't know what else I can describe beside my previous comment. Can you please clarify which part you would like to hear more about?

cottsay commented:
I did some research to try to understand why `__del__` is frowned upon. From what I gather, are you referring to the fact that it is hard to know if `__del__` will ever be called due to the way deallocation happens in Python?

dirk-thomas commented:
Yes, therefore for resource deallocation you should use an explicitly called function.

ablasdel commented:
Ok @dirk-thomas are you satisfied with the changes then? can we merge?

dirk-thomas commented on rqt_reconfigure/src/rqt_reconfigure/node_selector_widget.py:
```diff
@@ -398,16 +410,27 @@ def _add_children_treenode(self, treenodeitem_toplevel,
+    def _prune_nodetree_pernode(self):
+        """
+        """
+
```
The empty docblock should be removed.

dirk-thomas commented:
Beside the tiny comment the patch lgtm. I haven't tried to run it though.

ablasdel commented:
@cottsay once you pick that last nit and you feel this is ready please feel free to merge.

cottsay commented:
Okay, I think I've addressed everything. I found another small bug when a server disappears and fixed it. I think I've manually tried most of the prominent cases, so assuming Travis is happy, I'll probably merge this.

I tested by starting/stopping multiple DynRe servers with and without their entries selected in `rqt_reconfigure`, and using refresh to observe the correct behaviour in each case.

Thanks Dirk, for your constructive feedback!

cottsay commented:
Nevermind, looks like there is no Travis for rqt_common_plugins :)

ablasdel commented:
We all need our Travis! I think i'm going to start calling Dirk Travis
